Sources/X509/Extension Types/NameConstraints.swift

```diff
@@ -24,43 +24,266 @@ import SwiftASN1
 /// are consulted first, and if a name is matched in a forbidden tree then it does not matter whether
 /// the same name is also matched in a permitted tree.
 public struct NameConstraints {
-    /// The DNS name trees that are permitted in certificates issued by this CA.
-    ///
-    /// These restrictions are expressed in forms like `host.example.com`. Any DNS name that can be
-    /// constructed by adding zero or more labels to the left-hand side of the name satifies the constraint.
-    public var permittedDNSDomains: [String] {
-        // TODO(cory): We probably want a different collection type here that can lazily construct its contents, and
-        // that will let us support people assigning into this collection.
-        get {
-            self.permittedSubtrees.compactMap {
+    public struct DNSNames: Hashable, Sendable, Sequence, ExpressibleByArrayLiteral, CustomStringConvertible {
+        @inlinable
+        public static func ==(lhs: Self, rhs: Self) -> Bool {
+            lhs.elementsEqual(rhs)
+        }
+        
+        @inlinable
+        public func hash(into hasher: inout Hasher) {
+            hasher.combine(contentsOf: self)
+        }
+        
+        @inlinable
+        public var description: String {
+            "[\(self.joined(separator: ", "))]"
+        }
+        
+        @usableFromInline
+        var subtrees: [GeneralName]
+        
+        @inlinable
+        init(subtrees: [GeneralName]) {
+            self.subtrees = subtrees
+        }
+        
+        @inlinable
+        public init(_ elements: some Sequence<String>) {
+            self.subtrees = elements.map { .dnsName($0) }
+        }
+        
+        @inlinable
+        public init(arrayLiteral elements: String...) {
+            self.init(elements)
+        }
+        
+        @inlinable
+        public func makeIterator() -> some IteratorProtocol<String> {
+            self.subtrees.lazy.compactMap {
                 if case .dnsName(let name) = $0 {
                     return name
                 } else {
                     return nil
                 }
+            }.makeIterator()
```

Comment:
This is likely to perform extremely poorly: it should be sufficient to implement `Collection` over an opaque index type and use the free iterator.

Reply:
Why do you think this is expensive?

Reply:
The `.lazy` transforms are highly general, and their performance is often spotty.

Reply:
Adopted the `Collection` protocol with an opaque index